### PR TITLE
Add tag component and test

### LINKED ISSUE #2759: Tag component.
We need a component that can hold an arbitrary amount of "tags".
This would be convenient for construction (and other systems) as right now it uses components as tags. (See FirelockElectronicsComponent and GlassBeakerComponent)

### PR DESCRIPTION
**Pull request by DrSmugleaf**
Closes #2759 

This PR does not add it to any existing entities.

### CODE REVIEW

File: Content.Server/GameObjects/Components/Tag/TagComponent.cs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+using Robust.Shared.Serialization;
+
+namespace Content.Server.GameObjects.Components.Tag
+{
+    [RegisterComponent]
+    public class TagComponent : Component
+    {
+        public override string Name => "Tag";
+
+        private HashSet<string> _tags = new(0);
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Tag/TagComponent.cs:13]: I think using a 0 capacity is technically slower if the default capacity is already 0 and it doesn't save memory. pjb told me this about lists and I just checked and I'm pretty sure it applies to hashsets too.

File: Content.Server/GameObjects/Components/Tag/TagComponent.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
+﻿using System.Collections.Generic;
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Tag/TagComponent.cs:1]: plz nullable

**Review by metalgearsloth — COMMENTED** [review]
Using strings for it feels icky but I can't really think of anything better for a large number of tags.

File: Content.Server/GameObjects/Components/Tag/TagComponent.cs (hunk 1 above)
Comment (DrSmugleaf) [reply Content.Server/GameObjects/Components/Tag/TagComponent.cs]: Fixed.

File: Content.Server/GameObjects/Components/Tag/TagComponent.cs (hunk 2 above)
Comment (DrSmugleaf) [reply Content.Server/GameObjects/Components/Tag/TagComponent.cs:1]: For what sir
Should I make the methods accept nullable strings?

**Comment by metalgearsloth** [comment]
Sorry for stuffing you around on the pr, did we ever come up with an alternative to strings? e.g. having a hashset of enums? Keeping a master enum somewhere isn't the best but I'd prefer to avoid having hidden runtime issues due to typos (_cough Alien Colonial Marines_).

**Comment by PaulRitter** [comment]
is the sourcegen idea out the window? iirc it was pretty viable

**Review by Zumorica — CHANGES_REQUESTED** [review]
We discussed this and came to the conclusion that it'd be best to have tag prototypes.
Basically, we still use strings for tags, but also define the prototypes somewhere in YAML, just so we can index the prototypes when adding tags and printing a warning if a specific tag prototype doesn't exist! This way, we prevent all problems using strings has. (Typos. That's it.)
The new Tag prototype doesn't even need to do anything but exist and contain the tag string as its ID.

**Comment by DrSmugleaf** [comment]
Reviews addressed, changed tags to use prototypes.

**Review by Zumorica — APPROVED** [review]
LGTM, thank you very much!

**Review by metalgearsloth — APPROVED** [review]

**Comment by metalgearsloth** [comment]
I can die happy.

**Pull request merged**